File: ci/format_vote_gate.py
"""Format-specification vote gate (see `.github/workflows/format-vote-gate.yml`).

Structurally enforces the PMC vote required for Lance format-specification
changes (https://lance.org/community/voting/). The `format-change` label is
applied by the path labeler (`.github/labeler-area.yml`); this script reads it
and publishes the `format-spec-vote` commit status, which blocks merging until:

  * 3 PMC members have approved the PR (excluding the author), counted only on
    the head commit so new pushes invalidate stale approvals;
  * no PMC member has an outstanding "Request changes" review (a veto); and
  * the 72-hour voting period has elapsed. The clock starts once the PR is both
    labeled and out of draft, and pauses over weekends.

A PMC member can waive a trivial edit by applying the `format-waived` label.
Non-format PRs get a passing status immediately and are otherwise left alone.
Drafts get a blocking status but no comment: the vote has not opened yet.

The vote-counting and deadline rules are pure functions (`tally_reviews`,
`decide_verdict`, `vote_opened_at`, `weekday_deadline`) unit tested in
`test_format_vote_gate.py`; `main` wires them to the GitHub API.
"""

import json
import os
from collections import namedtuple
from datetime import datetime, timedelta, timezone
from zoneinfo import ZoneInfo

STATUS_CONTEXT = "format-spec-vote"
FORMAT_LABEL = "format-change"
WAIVED_LABEL = "format-waived"
COMMENT_MARKER = "<!-- format-spec-vote-status -->"
REQUIRED_APPROVALS = 3
PERIOD_HOURS = 72
VOTING_URL = "https://lance.org/community/voting/"

# The weekend boundary is fixed in UTC rather than a local zone: it has no DST
# transitions to reason about, and no PMC member's timezone gets to define when
# everyone else's clock pauses. Deadlines are *displayed* in UTC and Pacific.
WEEKEND_TZ = timezone.utc
DISPLAY_TZ = ZoneInfo("America/Los_Angeles")

# datetime.weekday() numbers Monday 0 .. Sunday 6, so the weekend is >= 5.
_SATURDAY = 5

# Review states that express a stance; COMMENTED/PENDING are ignored.
_STANCE_STATES = ("APPROVED", "CHANGES_REQUESTED", "DISMISSED")

TimelineFacts = namedtuple("TimelineFacts", "labeled_at waived ready_at")


def tally_reviews(reviews, head_sha, author, is_pmc):
    """Tally PMC votes from a PR's reviews.

    `reviews` is an ordered list of dicts with `login`, `state`, `commit_id`.
    A member's stance is their most recent stance review. Approvals only count
    on the head commit; earlier ones are stale. A "changes requested" review is
    a veto regardless of commit. The PR author never counts.
    """
    latest = {}
    for review in reviews:
        login = review["login"]
        if not login or not is_pmc(login) or login == author:
            continue
        if review["state"] not in _STANCE_STATES:
            continue
        latest[login.lower()] = review

    approvals, stale_approvals, vetoes = [], [], []
    for review in latest.values():
        if review["state"] == "APPROVED":
            target = approvals if review["commit_id"] == head_sha else stale_approvals
            target.append(review["login"])
        elif review["state"] == "CHANGES_REQUESTED":
            vetoes.append(review["login"])
    return approvals, stale_approvals, vetoes


def decide_verdict(veto_count, approval_count, period_elapsed, required):
    """Return the blocking condition (if any), in priority order."""
    if veto_count > 0:
        return "veto"
    if approval_count < required:
        return "insufficient"
    if not period_elapsed:
        return "waiting_period"
    return "pass"


def vote_opened_at(labeled_at, ready_at):
    """When the voting period starts, or None if it hasn't.

    A vote opens only once the proposal is both identified as a format change
    and offered for review, so the clock starts at the later of the two. A draft
    is still being drafted; time spent there shouldn't count toward the period.
    """
    if labeled_at is None or ready_at is None:
        return None
    return max(labeled_at, ready_at)


def _start_of_day(dt):
    return dt.replace(hour=0, minute=0, second=0, microsecond=0)


# Advance to Monday 00:00 if `dt` lands on a weekend; otherwise leave it alone.
def _skip_weekend(dt):
    while dt.weekday() >= _SATURDAY:
        dt = _start_of_day(dt) + timedelta(days=1)
    return dt


# Saturday 00:00 following `dt`, which must already be a weekday.
def _next_weekend(dt):
    return _start_of_day(dt) + timedelta(days=_SATURDAY - dt.weekday())


def weekday_deadline(start, hours):
    """When `hours` of non-weekend time have elapsed after `start`.

    Weekends don't count toward the voting period, so a proposal opened on a
    Friday afternoon doesn't burn most of its period while nobody is reading it.
    Both `start` and the result are aware datetimes; the arithmetic happens in
    `WEEKEND_TZ`, which decides where each weekend begins and ends.
    """
    cursor = _skip_weekend(start.astimezone(WEEKEND_TZ))
    remaining = timedelta(hours=hours)
    while True:
        until_weekend = _next_weekend(cursor) - cursor
        if remaining <= until_weekend:
            return cursor + remaining
        remaining -= until_weekend
        cursor = _skip_weekend(_next_weekend(cursor))


def _fmt_list(logins):
    return ", ".join(f"@{login}" for login in logins) if logins else "none"


# Renders as `Wed 2026-08-05 17:00 UTC (10:00 PDT)` — the PMC spans both zones.
# The Pacific weekday is spelled out only when the deadline falls on a different
# day there, which is the case that actually trips people up.
def _fmt_deadline(dt):
    local = dt.astimezone(DISPLAY_TZ)
    local_day = "" if local.date() == dt.date() else f"{local:%a }"
    return f"{dt:%a %Y-%m-%d %H:%M} UTC ({local_day}{local:%H:%M %Z})"


def _as_utc(dt):
    return dt.replace(tzinfo=timezone.utc) if dt.tzinfo is None else dt


def _build_comment(headline, approval_cell, vetoes, period_cell):
    return "\n".join(
        [
            COMMENT_MARKER,
            "> [!IMPORTANT]",
            "> ## Format specification vote",
            "",
            "This PR modifies the Lance format specification, so it requires "
            f"**{REQUIRED_APPROVALS} binding +1 votes from PMC members** "
            "(excluding the proposer) and a minimum "
            f"**{PERIOD_HOURS}-hour** voting period, weekends excluded, before "
            "it can merge. "
            "Vote by approving this PR (+1) or requesting changes (−1, a veto). "
            f"See the [voting process]({VOTING_URL}).",
            "",
            f"**Status: {headline}**",
            "",
            "| | |",
            "|---|---|",
            f"| Approvals (this commit) | {approval_cell} |",
            f"| Vetoes | {_fmt_list(vetoes)} |",
            f"| Voting period | {period_cell} |",
            "",
            "<sub>Updated automatically by the format-spec vote gate. A PMC member "
            f"may apply the `{WAIVED_LABEL}` label to waive the vote for a trivial "
            "edit (typo, wording, formatting).</sub>",
        ]
    )


def _load_pmc(workspace):
    import yaml

    roster_path = os.path.join(workspace, "docs", "src", "community", "pmc.yaml")
    with open(roster_path) as handle:
        roster = yaml.safe_load(handle)
    return {member["handle"].lower() for member in roster["members"]}


class Gate:
    def __init__(self, repo, pmc, run_url):
        self.repo = repo
        self.pmc = pmc
        self.run_url = run_url

    def is_pmc(self, login):
        return login is not None and login.lower() in self.pmc

    def set_status(self, sha, state, description):
        self.repo.get_commit(sha).create_status(
            state=state,
            context=STATUS_CONTEXT,
            description=description[:140],
            target_url=self.run_url,
        )

    def upsert_comment(self, issue, body):
        for comment in issue.get_comments():
            if COMMENT_MARKER in (comment.body or ""):
                if comment.body != body:
                    comment.edit(body)
                return
        issue.create_comment(body)

    def timeline_facts(self, issue):
        """Read the vote-clock inputs off the PR timeline in one pass."""
        labeled_at = None
        waived_by_pmc = False
        ready_at = None
        for event in issue.get_events():
            actor = event.actor.login if event.actor else None
            # A PR converted back to draft and re-opened for review restarts the
            # clock, so the *last* ready_for_review wins.
            if event.event == "ready_for_review":
                ready_at = _as_utc(event.created_at)
                continue
            if event.event != "labeled" or event.label is None:
                continue
            if event.label.name == FORMAT_LABEL and labeled_at is None:
                labeled_at = _as_utc(event.created_at)
            elif event.label.name == WAIVED_LABEL and self.is_pmc(actor):
                waived_by_pmc = True
        return TimelineFacts(labeled_at, waived_by_pmc, ready_at)

    def evaluate(self, number):
        pr = self.repo.get_pull(number)
        if pr.state != "open":
            print(f"PR #{number} is {pr.state}; skipping.")
            return
        head_sha = pr.head.sha
        labels = {label.name for label in pr.labels}

        # Non-format PRs get a passing status and are otherwise left alone.
        if FORMAT_LABEL not in labels:
            self.set_status(
                head_sha, "success", "No format-spec change; vote not required."
            )
            print(f"PR #{number}: not a format change.")
            return

        issue = self.repo.get_issue(number)
        facts = self.timeline_facts(issue)

        if WAIVED_LABEL in labels and facts.waived:
            self.set_status(
                head_sha, "success", "Format-spec vote waived by a PMC member."
            )
            print(f"PR #{number}: vote waived.")
            return

        # Stay quiet on drafts: the proposal isn't up for a vote yet, so there is
        # nothing for the PMC to act on and no deadline to announce.
        if pr.draft:
            self.set_status(
                head_sha,
                "failure",
                "Draft; voting period starts when marked ready for review.",
            )
            print(f"PR #{number}: draft, vote not open.")
            return

        reviews = [
            {
                "login": review.user.login if review.user else None,
                "state": review.state,
                "commit_id": review.commit_id,
            }
            for review in pr.get_reviews()
        ]
        approvals, stale, vetoes = tally_reviews(
            reviews, head_sha, pr.user.login, self.is_pmc
        )

        now = datetime.now(timezone.utc)
        # `pr.created_at` covers a PR opened ready for review, which never emits a
        # ready_for_review event.
        opened_at = vote_opened_at(
            facts.labeled_at, facts.ready_at or _as_utc(pr.created_at)
        )
        period_ends = weekday_deadline(opened_at or now, PERIOD_HOURS)
        period_elapsed = now >= period_ends
        verdict = decide_verdict(
            len(vetoes), len(approvals), period_elapsed, REQUIRED_APPROVALS
        )

        deadline = _fmt_deadline(period_ends)
        if verdict == "veto":
            state, summary = "failure", f"Vetoed by {len(vetoes)} PMC member(s)."
            headline = f"❌ Blocked — vetoed by {_fmt_list(vetoes)}"
        elif verdict == "insufficient":
            state = "failure"
            summary = (
                f"{len(approvals)}/{REQUIRED_APPROVALS} PMC approvals on this commit."
            )
            headline = f"❌ Blocked — {len(approvals)} of {REQUIRED_APPROVALS} required approvals"
        elif verdict == "waiting_period":
            state, summary = "failure", f"Approved; voting period ends {deadline}."
            headline = (
                f"⏳ Approvals met ({len(approvals)}/{REQUIRED_APPROVALS}); "
                f"voting period ends {deadline}"
            )
        else:
            state = "success"
            summary = f"Passed — {len(approvals)} PMC approvals, period elapsed."
            headline = f"✅ Vote passed — {len(approvals)} PMC approvals, voting period elapsed"

        period_cell = (
            f"elapsed — ended {deadline}" if period_elapsed else f"ends {deadline}"
        )
        approval_cell = (
            f"{_fmt_list(approvals)} ({len(approvals)}/{REQUIRED_APPROVALS})"
        )
        if stale:
            approval_cell += f" — stale, re-approve needed: {_fmt_list(stale)}"

        self.set_status(head_sha, state, summary)
        self.upsert_comment(
            issue, _build_comment(headline, approval_cell, vetoes, period_cell)
        )
        print(f"PR #{number}: {summary}")


def main():
    from github import Github

    workspace = os.environ["GITHUB_WORKSPACE"]
    token = os.environ["GITHUB_TOKEN"]
    repo_name = os.environ["GITHUB_REPOSITORY"]
    event_name = os.environ["GITHUB_EVENT_NAME"]
    run_url = (
        f"{os.environ['GITHUB_SERVER_URL']}/{repo_name}/actions/runs/"
        f"{os.environ['GITHUB_RUN_ID']}"
    )

    repo = Github(token).get_repo(repo_name)
    gate = Gate(repo, _load_pmc(workspace), run_url)

    if event_name == "schedule":
        # The schedule trigger has no PR context, so sweep every open
        # format-change PR to re-check the voting-period clock.
        pulls = [
            pr
            for pr in repo.get_pulls(state="open")
            if any(label.name == FORMAT_LABEL for label in pr.labels)
        ]
        print(f"Scheduled sweep: {len(pulls)} open {FORMAT_LABEL} PR(s).")
        for pr in pulls:
            try:
                gate.evaluate(pr.number)
            except Exception as err:  # noqa: BLE001 - keep sweeping other PRs
                print(f"PR #{pr.number}: {err}")
    else:
        with open(os.environ["GITHUB_EVENT_PATH"]) as handle:
            event = json.load(handle)
        gate.evaluate(event["pull_request"]["number"])


if __name__ == "__main__":
    main()

File: ci/format_vote_gate.js
// Logic for the format-specification vote gate
// (`.github/workflows/format-vote-gate.yml`). Extracted into a module so the
// vote-counting rules can be unit tested (see `format_vote_gate.test.js`); the
// workflow loads it from the trusted base checkout and calls `run`.

const fs = require("fs");
const path = require("path");

const STATUS_CONTEXT = "format-spec-vote";
const LABEL = "format-change";
const COMMENT_MARKER = "<!-- format-spec-vote-status -->";
const REQUIRED_APPROVALS = 3;
const PERIOD_DAYS = 7;
const PERIOD_MS = PERIOD_DAYS * 24 * 60 * 60 * 1000;
const VOTING_URL = "https://lance.org/community/voting/";
const DAY_MS = 24 * 60 * 60 * 1000;

// The Lance format specification is the proto definitions and the spec docs.
// Changes to either require a PMC vote.
function isFormatFile(p) {
  return (
    (p.endsWith(".proto") && p.startsWith("protos/")) ||
    p.startsWith("docs/src/format/")
  );
}

// Each PMC member's current stance is their most recent non-comment review.
// Approvals are only counted on the head commit (stale approvals from earlier
// commits do not count toward the bar); a "changes requested" review is a veto
// and blocks regardless of commit until withdrawn. The PR author never counts,
// even if they are on the PMC.
function tallyReviews(reviews, headSha, author, isPmc) {
  const latest = new Map();
  for (const r of reviews) {
    const login = r.user && r.user.login;
    if (!login || !isPmc(login) || login === author) continue;
    if (!["APPROVED", "CHANGES_REQUESTED", "DISMISSED"].includes(r.state)) {
      continue;
    }
    latest.set(login.toLowerCase(), {
      login,
      state: r.state,
      commitId: r.commit_id,
    });
  }
  const approvals = [];
  const staleApprovals = [];
  const vetoes = [];
  for (const v of latest.values()) {
    if (v.state === "APPROVED") {
      (v.commitId === headSha ? approvals : staleApprovals).push(v.login);
    } else if (v.state === "CHANGES_REQUESTED") {
      vetoes.push(v.login);
    }
  }
  return { approvals, staleApprovals, vetoes };
}

// Pure decision: which blocking condition (if any) applies, in priority order.
function decideVerdict({ vetoCount, approvalCount, periodElapsed, required }) {
  if (vetoCount > 0) return "veto";
  if (approvalCount < required) return "insufficient";
  if (!periodElapsed) return "waiting_period";
  return "pass";
}

const fmtList = (xs) => (xs.length ? xs.map((u) => `@${u}`).join(", ") : "none");

function loadPmc(workspace) {
  const yaml = require(path.join(workspace, "node_modules", "js-yaml"));
  const rosterPath = path.join(workspace, "docs", "src", "community", "pmc.yaml");
  const roster = yaml.load(fs.readFileSync(rosterPath, "utf8"));
  return new Set(roster.members.map((m) => m.handle.toLowerCase()));
}

async function run({ github, context, core }) {
  const { owner, repo } = context.repo;
  const workspace = process.env.GITHUB_WORKSPACE;
  const pmc = loadPmc(workspace);
  const isPmc = (login) => login != null && pmc.has(login.toLowerCase());
  const runUrl = `${context.serverUrl}/${owner}/${repo}/actions/runs/${context.runId}`;

  async function setStatus(sha, state, description) {
    await github.rest.repos.createCommitStatus({
      owner,
      repo,
      sha,
      state,
      context: STATUS_CONTEXT,
      description: description.slice(0, 140),
      target_url: runUrl,
    });
  }

  async function upsertComment(prNumber, body) {
    const full = `${COMMENT_MARKER}\n${body}`;
    const comments = await github.paginate(github.rest.issues.listComments, {
      owner,
      repo,
      issue_number: prNumber,
      per_page: 100,
    });
    const existing = comments.find((c) => c.body && c.body.includes(COMMENT_MARKER));
    if (existing) {
      if (existing.body !== full) {
        await github.rest.issues.updateComment({
          owner,
          repo,
          comment_id: existing.id,
          body: full,
        });
      }
    } else {
      await github.rest.issues.createComment({
        owner,
        repo,
        issue_number: prNumber,
        body: full,
      });
    }
  }

  // Returns { firstAddedAt, removedByPmc } for LABEL on this PR, from the
  // timeline: when the label was first applied (the vote-period anchor) and
  // whether a PMC member deliberately removed it (a trivial-edit waiver).
  async function labelHistory(prNumber) {
    const events = await github.paginate(
      github.rest.issues.listEventsForTimeline,
      { owner, repo, issue_number: prNumber, per_page: 100 },
    );
    let firstAddedAt = null;
    let removedByPmc = false;
    for (const e of events) {
      if (!e.label || e.label.name !== LABEL) continue;
      if (e.event === "labeled" && !firstAddedAt) {
        firstAddedAt = e.created_at;
      } else if (e.event === "unlabeled" && isPmc(e.actor && e.actor.login)) {
        removedByPmc = true;
      }
    }
    return { firstAddedAt, removedByPmc };
  }

  async function evaluate(prNumber) {
    const { data: pr } = await github.rest.pulls.get({
      owner,
      repo,
      pull_number: prNumber,
    });
    if (pr.state !== "open") {
      core.info(`PR #${prNumber} is ${pr.state}; skipping.`);
      return;
    }
    const headSha = pr.head.sha;
    const author = pr.user.login;

    const files = await github.paginate(github.rest.pulls.listFiles, {
      owner,
      repo,
      pull_number: prNumber,
      per_page: 100,
    });
    const touchesFormat = files.map((f) => f.filename).some(isFormatFile);

    let hasLabel = pr.labels.some((l) => l.name === LABEL);
    const { firstAddedAt, removedByPmc } = await labelHistory(prNumber);

    // Auto-apply the label, unless a PMC member deliberately removed it.
    if (touchesFormat && !hasLabel && !removedByPmc) {
      await github.rest.issues.addLabels({
        owner,
        repo,
        issue_number: prNumber,
        labels: [LABEL],
      });
      hasLabel = true;
    }

    if (!hasLabel) {
      await setStatus(
        headSha,
        "success",
        removedByPmc
          ? "Format-spec vote waived by a PMC member."
          : "No format-spec change; vote not required.",
      );
      if (removedByPmc) {
        await upsertComment(
          prNumber,
          `> [!NOTE]\n> A PMC member removed the \`${LABEL}\` label, ` +
            `waiving the format-spec vote for this PR (trivial edit).`,
        );
      }
      core.info(`PR #${prNumber}: no vote required.`);
      return;
    }

    const reviews = await github.paginate(github.rest.pulls.listReviews, {
      owner,
      repo,
      pull_number: prNumber,
      per_page: 100,
    });
    const { approvals, staleApprovals, vetoes } = tallyReviews(
      reviews,
      headSha,
      author,
      isPmc,
    );

    const now = new Date();
    const voteStart = firstAddedAt ? new Date(firstAddedAt) : now;
    const periodEnds = new Date(voteStart.getTime() + PERIOD_MS);
    const periodElapsed = now >= periodEnds;
    const verdict = decideVerdict({
      vetoCount: vetoes.length,
      approvalCount: approvals.length,
      periodElapsed,
      required: REQUIRED_APPROVALS,
    });

    const endDate = periodEnds.toISOString().slice(0, 10);
    let state, summary, headline;
    if (verdict === "veto") {
      state = "failure";
      summary = `Vetoed by ${vetoes.length} PMC member(s).`;
      headline = `❌ Blocked — vetoed by ${fmtList(vetoes)}`;
    } else if (verdict === "insufficient") {
      state = "failure";
      summary = `${approvals.length}/${REQUIRED_APPROVALS} PMC approvals on this commit.`;
      headline = `❌ Blocked — ${approvals.length} of ${REQUIRED_APPROVALS} required approvals`;
    } else if (verdict === "waiting_period") {
      state = "failure";
      summary = `Approved; voting period ends ${endDate}.`;
      headline =
        `⏳ Approvals met (${approvals.length}/${REQUIRED_APPROVALS}); ` +
        `voting period ends ${endDate}`;
    } else {
      state = "success";
      summary = `Passed — ${approvals.length} PMC approvals, period elapsed.`;
      headline = `✅ Vote passed — ${approvals.length} PMC approvals, voting period elapsed`;
    }

    const daysLeft = Math.max(0, Math.ceil((periodEnds.getTime() - now.getTime()) / DAY_MS));
    const periodCell = periodElapsed
      ? `elapsed (ended ${endDate})`
      : `ends ${endDate} (${daysLeft} day(s) left)`;
    const approvalCell =
      `${fmtList(approvals)} (${approvals.length}/${REQUIRED_APPROVALS})` +
      (staleApprovals.length
        ? ` — stale, re-approve needed: ${fmtList(staleApprovals)}`
        : "");

    const body = [
      "> [!IMPORTANT]",
      "> ## Format specification vote",
      "",
      "This PR modifies the Lance format specification, so it requires " +
        `**${REQUIRED_APPROVALS} binding +1 votes from PMC members** ` +
        "(excluding the proposer) and a minimum " +
        `**${PERIOD_DAYS}-day** voting period before it can merge. ` +
        "Vote by approving this PR (+1) or requesting changes (−1, a veto). " +
        `See the [voting process](${VOTING_URL}).`,
      "",
      `**Status: ${headline}**`,
      "",
      "| | |",
      "|---|---|",
      `| Approvals (this commit) | ${approvalCell} |`,
      `| Vetoes | ${fmtList(vetoes)} |`,
      `| Voting period | ${periodCell} |`,
      "",
      "<sub>Updated automatically by the format-spec vote gate. A PMC member " +
        `may remove the \`${LABEL}\` label to waive the vote for a trivial edit ` +
        "(typo, wording, formatting).</sub>",
    ].join("\n");

    await setStatus(headSha, state, summary);
    await upsertComment(prNumber, body);
    core.info(`PR #${prNumber}: ${summary}`);
  }

  if (context.eventName === "schedule") {
    const issues = await github.paginate(github.rest.issues.listForRepo, {
      owner,
      repo,
      state: "open",
      labels: LABEL,
      per_page: 100,
    });
    const pulls = issues.filter((i) => i.pull_request);
    core.info(`Scheduled sweep: ${pulls.length} open ${LABEL} PR(s).`);
    for (const p of pulls) {
      try {
        await evaluate(p.number);
      } catch (err) {
        core.warning(`PR #${p.number}: ${err.message}`);
      }
    }
  } else {
    await evaluate(context.payload.pull_request.number);
  }
}

module.exports = {
  run,
  isFormatFile,
  tallyReviews,
  decideVerdict,
  constants: { STATUS_CONTEXT, LABEL, REQUIRED_APPROVALS, PERIOD_DAYS },
};
